### ISSUE TITLE
blas_kernels.cu : fill_ongpu() CUDA Error: invalid device function: No error

### ISSUE THREAD

**Issue by carloSpa-ET** [AlexeyAB/darknet#4414]
Hi there, i am facing some strange issues with the training function.
I have compiled the source with CMake and checked cuda and cudnn, everything compile successful,
I have created my own Dataset, with images and label created as you have described in your tutorial.
I need to train a yolov3 with 2 classes, so the following is my cfg file:
```
batch=64
subdivisions=64
width=320
height=320
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 4000
policy=steps
steps=3200,3600
scales=.1,.1

...

[convolutional]
size=1
stride=1
pad=1
filters=21
activation=linear

[yolo]
mask = 6,7,8
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=2
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1


[route]
layers = -4

...

[convolutional]
size=1
stride=1
pad=1
filters=21
activation=linear


[yolo]
mask = 3,4,5
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=2
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

...

[convolutional]
size=1
stride=1
pad=1
filters=21
activation=linear


[yolo]
mask = 0,1,2
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=2
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1 
```

Then i have downloaded the file **darknet53.conv.74**
and executing the command:
`./darknet.exe detector train data/obj.data cfg/yolo-obj.cfg darknet53.conv.74`
Then this happen:
```
yolo-obj
net.optimized_memory = 0
batch = 1, time_steps = 1, train = 1
   layer   filters  size/strd(dil)      input                output
   0 conv     32       3 x 3/ 1    320 x 320 x   3 ->  320 x 320 x  32 0.177 BF
   1 conv     64       3 x 3/ 2    320 x 320 x  32 ->  160 x 160 x  64 0.944 BF
   2 conv     32       1 x 1/ 1    160 x 160 x  64 ->  160 x 160 x  32 0.105 BF
   3 conv     64       3 x 3/ 1    160 x 160 x  32 ->  160 x 160 x  64 0.944 BF
   4 Shortcut Layer: 1
   5 conv    128       3 x 3/ 2    160 x 160 x  64 ->   80 x  80 x 128 0.944 BF
   6 conv     64       1 x 1/ 1     80 x  80 x 128 ->   80 x  80 x  64 0.105 BF
   7 conv    128       3 x 3/ 1     80 x  80 x  64 ->   80 x  80 x 128 0.944 BF
   8 Shortcut Layer: 5
   9 conv     64       1 x 1/ 1     80 x  80 x 128 ->   80 x  80 x  64 0.105 BF
  10 conv    128       3 x 3/ 1     80 x  80 x  64 ->   80 x  80 x 128 0.944 BF
  11 Shortcut Layer: 8
  12 conv    256       3 x 3/ 2     80 x  80 x 128 ->   40 x  40 x 256 0.944 BF
  13 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  14 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  15 Shortcut Layer: 12
  16 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  17 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  18 Shortcut Layer: 15
  19 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  20 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  21 Shortcut Layer: 18
  22 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  23 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  24 Shortcut Layer: 21
  25 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  26 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  27 Shortcut Layer: 24
  28 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  29 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  30 Shortcut Layer: 27
  31 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  32 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  33 Shortcut Layer: 30
  34 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
  35 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
  36 Shortcut Layer: 33
  37 conv    512       3 x 3/ 2     40 x  40 x 256 ->   20 x  20 x 512 0.944 BF
  38 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  39 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  40 Shortcut Layer: 37
  41 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  42 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  43 Shortcut Layer: 40
  44 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  45 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  46 Shortcut Layer: 43
  47 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  48 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  49 Shortcut Layer: 46
  50 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  51 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  52 Shortcut Layer: 49
  53 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  54 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  55 Shortcut Layer: 52
  56 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  57 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  58 Shortcut Layer: 55
  59 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  60 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  61 Shortcut Layer: 58
  62 conv   1024       3 x 3/ 2     20 x  20 x 512 ->   10 x  10 x1024 0.944 BF
  63 conv    512       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x 512 0.105 BF
  64 conv   1024       3 x 3/ 1     10 x  10 x 512 ->   10 x  10 x1024 0.944 BF
  65 Shortcut Layer: 62
  66 conv    512       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x 512 0.105 BF
  67 conv   1024       3 x 3/ 1     10 x  10 x 512 ->   10 x  10 x1024 0.944 BF
  68 Shortcut Layer: 65
  69 conv    512       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x 512 0.105 BF
  70 conv   1024       3 x 3/ 1     10 x  10 x 512 ->   10 x  10 x1024 0.944 BF
  71 Shortcut Layer: 68
  72 conv    512       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x 512 0.105 BF
  73 conv   1024       3 x 3/ 1     10 x  10 x 512 ->   10 x  10 x1024 0.944 BF
  74 Shortcut Layer: 71
  75 conv    512       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x 512 0.105 BF
  76 conv   1024       3 x 3/ 1     10 x  10 x 512 ->   10 x  10 x1024 0.944 BF
  77 conv    512       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x 512 0.105 BF
  78 conv   1024       3 x 3/ 1     10 x  10 x 512 ->   10 x  10 x1024 0.944 BF
  79 conv    512       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x 512 0.105 BF
  80 conv   1024       3 x 3/ 1     10 x  10 x 512 ->   10 x  10 x1024 0.944 BF
  81 conv     21       1 x 1/ 1     10 x  10 x1024 ->   10 x  10 x  21 0.004 BF
  82 yolo
[yolo] params: iou loss: mse (2), iou_norm: 0.75, cls_norm: 1.00, scale_x_y: 1.00
  83 route  79                                     ->   10 x  10 x 512
  84 conv    256       1 x 1/ 1     10 x  10 x 512 ->   10 x  10 x 256 0.026 BF
  85 upsample                 2x    10 x  10 x 256 ->   20 x  20 x 256
  86 route  85 61                                  ->   20 x  20 x 768
  87 conv    256       1 x 1/ 1     20 x  20 x 768 ->   20 x  20 x 256 0.157 BF
  88 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  89 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  90 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  91 conv    256       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x 256 0.105 BF
  92 conv    512       3 x 3/ 1     20 x  20 x 256 ->   20 x  20 x 512 0.944 BF
  93 conv     21       1 x 1/ 1     20 x  20 x 512 ->   20 x  20 x  21 0.009 BF
  94 yolo
[yolo] params: iou loss: mse (2), iou_norm: 0.75, cls_norm: 1.00, scale_x_y: 1.00
  95 route  91                                     ->   20 x  20 x 256
  96 conv    128       1 x 1/ 1     20 x  20 x 256 ->   20 x  20 x 128 0.026 BF
  97 upsample                 2x    20 x  20 x 128 ->   40 x  40 x 128
  98 route  97 36                                  ->   40 x  40 x 384
  99 conv    128       1 x 1/ 1     40 x  40 x 384 ->   40 x  40 x 128 0.157 BF
 100 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
 101 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
 102 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
 103 conv    128       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x 128 0.105 BF
 104 conv    256       3 x 3/ 1     40 x  40 x 128 ->   40 x  40 x 256 0.944 BF
 105 conv     21       1 x 1/ 1     40 x  40 x 256 ->   40 x  40 x  21 0.017 BF
 106 yolo
[yolo] params: iou loss: mse (2), iou_norm: 0.75, cls_norm: 1.00, scale_x_y: 1.00
Total BFLOPS 38.637
 Allocate additional workspace_size = 52.43 MB
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
Resizing
480 x 480
 try to allocate additional workspace_size = 52.43 MB
 CUDA allocate done!
Loaded: 0.000000 seconds
CUDA status Error: file: D:/Yolov3Test/Yolov3/darknet-master/src/blas_kernels.cu : fill_ongpu() : line: 668 : build time: Nov 28 2019 - 13:53:10
CUDA Error: invalid device function
CUDA Error: invalid device function: No error
```

i can't really undertand what this error mean, i have searched for days in your issues in git's issues, but i can't find anything about it..
I have a gtx 1070 with 8GB
CUDA 10.1
python 3.7
opencv compiled with cuda support 3.4.8
Windows 10
all of my drivers are updated

if it can help it stop working at this function:
network_kernel.cu

```
void forward_network_gpu(network net, network_state state)
{
    state.workspace = net.workspace;
    int i;
    for(i = 0; i < net.n; ++i){
        state.index = i;
        layer l = net.layers[i];
        if(l.delta_gpu && state.train){
            fill_ongpu(l.outputs * l.batch, 0, l.delta_gpu, 1);  //<------ this call cause the crash
        }
        l.forward_gpu(l, state);
        if(net.wait_stream)
            cudaStreamSynchronize(get_cuda_stream());
        state.input = l.output_gpu;
    }
}
 ```

Thanks in advance.

**Comment by AlexeyAB** [comment]
It seems Cmake uses incorrect ARCH for your GPU, or you use several old CUDA versions.

Try to build with `make` rather than `cmake`: https://github.com/AlexeyAB/darknet#how-to-compile-on-linux-using-make

set GPU=1 CUDNN=1 OPENCV=1 in the Makefile

**Comment by carloSpa-ET** [comment]
hi alexey, i can't take screenshots right now, but my CUDA folder contains 2 directories, one is an old installation of Cuda 10.0 and the other one is a fresh installation of Cuda 10.1, i have checked the cmake gui and the system is using the folder v10.1,
CUDA_ARCHITECTURES is set to auto, and it use version 6.1 that as far as i know is the architecture of my GPU.

i am really sorry that i can't post the screenshot, i will post them Monday as soon as i can, thank you.

**Comment by carloSpa-ET** [comment]
hi @AlexeyAB, 
I've take the screenshots you have asked for:

CudaFolder:
![CudaFolder](https://user-images.githubusercontent.com/58331649/69944005-8125e780-14e6-11ea-9e0e-454d34ca96a7.png)

CMake:
![CMake](https://user-images.githubusercontent.com/58331649/69944010-85520500-14e6-11ea-8ed1-8d8e01bdd7f0.png)

**Comment by AlexeyAB** [comment]
So it doesn't work if you recompile with CUDA_ARCHITECTURES 3.0 7.5 ?

**Comment by carloSpa-ET** [comment]
nope, but i've tried to use another version of opencv, cause i was thinking that it wasn't using my version of opencv compiled with cuda, and after switching opencv it started throwing the error that it can't find the opencv's DLLs and PThreads.dll, so i pasted PTHreads.dll from 3rdparty folder of darknet, and all the DLLs from bin folder of the new opencv that i previously built, and now it works!
But is it ok that i have to manually copy all of the opencv's DLLs from the origin folder to the project folder?
Thanks

**Comment by AlexeyAB** [comment]
> But is it ok that i have to manually copy all of the opencv's DLLs from the origin folder to the project folder?

Yes.

Another way to set paths to OpenCV dll's in the Window system variable `PATH`  https://en.wikipedia.org/wiki/PATH_(variable)#DOS,_OS/2,_and_Windows

![edit-environment-variables](https://user-images.githubusercontent.com/4096485/69957886-b3e6d480-1514-11ea-9450-93bd4570e907.png)

**Comment by yangdashi888** [comment]
the problem is beacause of your CUDA10.0。when you cmake the project, CMAKE autoly add the part of CUDA10.0 to you project. 
Examples are as follows:
![360截图20200304111503215](https://user-images.githubusercontent.com/20290902/75842810-5f513f00-5e0c-11ea-8ccd-0da6d794c3f8.jpg)
The second can use the CUDA10.0，but the build project doesn't go wrong. When use the exe,
you will get an error of "fill_ongpu() CUDA Error: invalid device function: No error "
The solution is to remove the environment Variable of "CUDA_PATH_V10_1".

**Comment by AlexeyAB** [comment]
@yangdashi888 

> the problem is beacause of your CUDA10.0。when you cmake the project, CMAKE autoly add the part of CUDA10.0 to you project.
> The second can use the CUDA10.0，but the build project doesn't go wrong.

I don't understand. What is wrong with CUDA10.0?
